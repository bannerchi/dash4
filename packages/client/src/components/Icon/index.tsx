import React from 'react';
import withStyles, { WithStyles } from 'react-jss';

const styles = {
	root: {
		position: 'relative',
		display: 'block',
	},
	'size-s': {
		width: 14,
		height: 14,
	},
	'size-m': {
		width: 18,
		height: 18,
	},
	'size-l': {
		width: 24,
		height: 24,
	},
	'size-i-s': {
		fontSize: 14,
	},
	'size-i-m': {
		fontSize: 18,
	},
	'size-i-l': {
		fontSize: 24,
	},
};

export const sizeMap = {
	s: 's' as 's',
	m: 'm' as 'm',
	l: 'l' as 'l',
};

export type TSize = keyof typeof sizeMap;

export const colorMap = {
	dark: 'dark' as 'dark',
	light: 'light' as 'light',
};

export type TColor = keyof typeof colorMap;

interface IProps extends WithStyles<typeof styles> {
	name: string;
	size?: TSize;
	color?: TColor;
}

export const Icon = withStyles(styles)(({ classes, name, size }: IProps) => {
	const classNames: string[] = [classes.root];
	const classNamesI: string[] = ['material-icons'];
	if (size) {
		classNames.push(classes[`size-${size}`]);
		classNamesI.push(classes[`size-i-${size}`]);
	}
	return (
		<span className={classNames.join(' ')}>
			<i className={classNamesI.join(' ')}>{name}</i>
		</span>
	);
});
