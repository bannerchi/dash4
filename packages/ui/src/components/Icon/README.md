<div align="center">
<h1>Dash4 Icon</h1>

Render a [material ui](https://material.io/tools/icons/?style=baseline) icon
<br />

</div>

<!-- STORY -->

```tsx
import React from 'react';
import { Icon } from '@dash4/ui';

export const YourComponent = () => {
	return <Icon name="play_arrow" size="m" />;
};
```

_props:_

```ts
// icon name defined by material ui icons
name: string;
className?: string;
// type TSize = "s" | "m" | "l"
size?: TSize;
// type TColor = "dark" | "light"
color?: TColor;
// type TAnimation = "rotation-clockwise" | "rotation-counter-clockwise"
animation?: TAnimation;
// type TAlign = "center-in-content"
align?: TAlign;
// type TDisplay = "inline" | "block" (default=block)
display?: TDisplay;
```
