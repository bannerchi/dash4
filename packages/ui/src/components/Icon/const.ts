export const sizeMap = {
	s: 's',
	m: 'm',
	l: 'l',
} as const;

export type TSize = keyof typeof sizeMap;

export const displayMap = {
	inline: 'inline',
	block: 'block',
} as const;

export type TDisplay = keyof typeof displayMap;

export const colorMap = {
	dark: 'dark',
	bright: 'bright',
} as const;

export type TColor = keyof typeof colorMap;

export const alignMap = {
	'center-in-content': 'center-in-content',
} as const;

export type TAlign = keyof typeof alignMap;

export const animationMap = {
	'rotation-clockwise': 'rotation-clockwise',
	'rotation-counter-clockwise': 'rotation-counter-clockwise',
} as const;

export type TAnimation = keyof typeof animationMap;
