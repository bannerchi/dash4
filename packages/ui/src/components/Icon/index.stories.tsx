/// <reference types="../../types/markdown" />
import { storiesOf } from '@storybook/react';
import React from 'react';
import { alignMap, animationMap, colorMap, Icon, sizeMap, TAlign, TAnimation, TColor, TSize } from '.';
import Readme from './README.md';

const stories = storiesOf('Icon', module);
stories.add(
	'docs',
	() => (
		<div style={{ textAlign: 'center' }}>
			<Icon name="play_arrow" />
			<Icon name="favorite" />
			<Icon name="bookmark_border" />
		</div>
	),
	{
		readme: {
			content: Readme,
		},
	}
);
stories.add('default', () => <Icon name="play_arrow" />);
Object.keys(sizeMap).forEach((size) => {
	stories.add(`size ${size}`, () => <Icon name="play_arrow" size={size as TSize} />);
});
Object.keys(colorMap).forEach((color) => {
	stories.add(`color ${color}`, () => (
		<>
			<style>{`body { background: #ccc; }`}</style>
			<Icon name="play_arrow" color={color as TColor} />
		</>
	));
});
Object.keys(alignMap).forEach((align) => {
	stories.add(`align ${align}`, () => (
		<div style={{ width: '100vw', height: '100vh' }}>
			<Icon name="play_arrow" align={align as TAlign} />
		</div>
	));
});
Object.keys(animationMap).forEach((animation) => {
	stories.add(`animation ${animation}`, () => <Icon name="cached" animation={animation as TAnimation} />);
});
