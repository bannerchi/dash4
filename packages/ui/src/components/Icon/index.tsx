/** @jsx jsx */
import { css, jsx, keyframes } from '@emotion/core';
import { TAlign, TAnimation, TColor, TDisplay, TSize } from './const';

interface IProps {
	// icon name defined by [material ui icons](https://material.io/tools/icons/?style=baseline)
	name: string;
	className?: string;
	// type TSize = "s" | "m" | "l"
	size?: TSize;
	// type TColor = "dark" | "bright"
	color?: TColor;
	// type TAnimation = "rotation-clockwise" | "rotation-counter-clockwise"
	animation?: TAnimation;
	// type TAlign = "center-in-content"
	align?: TAlign;
	// true = display inline
	inline?: boolean;
}

function getSize(size: TSize) {
	if (size === 'l') {
		return 24;
	}
	if (size === 'm') {
		return 18;
	}
	return 14;
}

const rotationClockwise = keyframes`
	from {
		transform: rotate(0deg);
	}
	to {
		transform: rotate(359deg);
	}
`;
const rotationCounterClockwise = keyframes`
	from {
		transform: rotate(359deg);
	}
	to {
		transform: rotate(0deg);
	}
`;

export const Icon = ({ className = '', color, name, size, animation, align, inline }: IProps) => {
	return (
		<span
			className={className}
			css={[
				css`
					position: relative;
					display: ${inline ? 'inline' : 'block'};
				`,
				color === 'bright' &&
					css`
						color: #fff;
					`,
				color === 'dark' &&
					css`
						color: #000;
					`,
				size &&
					css`
						width: ${getSize(size)}px;
						height: ${getSize(size)}px;
					`,
				align === 'center-in-content' &&
					css`
						position: absolute;
						top: 0;
						left: 0;
						height: 100%;
						width: 100%;
						display: flex;
						justify-content: center;
						align-items: center;
						&::before {
							content: ' ';
							display: block;
							padding-top: 100% /* initial ratio of 1:1*/;
						}
					`,
			]}
		>
			<i
				className="material-icons"
				css={[
					size &&
						css`
							font-size: ${getSize(size)}px;
						`,
					!size &&
						css`
							font-size: inherit;
						`,
					animation === 'rotation-clockwise' &&
						css`
							animation: ${rotationClockwise} 1.5s infinite linear;
						`,
					animation === 'rotation-counter-clockwise' &&
						css`
							animation: ${rotationCounterClockwise} 1.5s infinite linear;
						`,
				]}
			>
				{name}
			</i>
		</span>
	);
};

export * from './const';
