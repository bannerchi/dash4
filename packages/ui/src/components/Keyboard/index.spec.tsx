/// <reference types="@types/jest" />
import { render } from '@testing-library/react';
import React from 'react';
import { Icon } from '../Icon';
import { Key, Keyboard } from './index';

describe('Keyboard', () => {
	test('exist', () => {
		expect(typeof Keyboard).toBe('function');
	});
	test('render default', () => {
		const { container } = render(
			<Keyboard>
				<Key>A</Key>
			</Keyboard>
		);
		expect(container.firstChild).toMatchSnapshot();
	});
	test('render with className', () => {
		const { container } = render(
			<Keyboard className="test">
				<Key>A</Key>
			</Keyboard>
		);
		expect(container.firstChild).toMatchSnapshot();
	});
	test('render with Icon', () => {
		const { container } = render(
			<Keyboard className="test">
				<Key>A</Key>
				<Key className="test">
					<Icon name="arrow_upward" />
				</Key>
			</Keyboard>
		);
		expect(container.firstChild).toMatchSnapshot();
	});
	test('Key standalonw', () => {
		const { container } = render(<Key>A</Key>);
		expect(container.firstChild).toMatchSnapshot();
	});
	test('Key with custom tagname', () => {
		const { container } = render(<Key tagName="div">A</Key>);
		expect(container.firstChild).toMatchSnapshot();
	});
});
