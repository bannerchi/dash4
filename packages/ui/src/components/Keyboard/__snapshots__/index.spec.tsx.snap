// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard Key standalonw 1`] = `
.emotion-0 {
  position: relative;
  display: inline;
  padding: 2px 5px;
  background-image: linear-gradient(rgb(223,224,231),rgb(248,249,255));
  border-radius: 4px;
  box-shadow: 0px 0px 1px 1px rgba(112,113,117,0.8);
  font-size: 12px;
  line-height: 12px;
  color: rgb(152,153,157);
}

.emotion-0 + .emotion-0 {
  margin-left: 18px;
}

.emotion-0 + .emotion-0:before {
  content: '+';
  position: absolute;
  left: -18px;
  top: 0;
  padding: 2px 4px;
  color: rgb(112,113,117);
  font-size: 16px;
}

<span
  class=" emotion-0"
>
  A
</span>
`;

exports[`Keyboard Key with custom tagname 1`] = `
.emotion-0 {
  position: relative;
  display: inline;
  padding: 2px 5px;
  background-image: linear-gradient(rgb(223,224,231),rgb(248,249,255));
  border-radius: 4px;
  box-shadow: 0px 0px 1px 1px rgba(112,113,117,0.8);
  font-size: 12px;
  line-height: 12px;
  color: rgb(152,153,157);
}

.emotion-0 + .emotion-0 {
  margin-left: 18px;
}

.emotion-0 + .emotion-0:before {
  content: '+';
  position: absolute;
  left: -18px;
  top: 0;
  padding: 2px 4px;
  color: rgb(112,113,117);
  font-size: 16px;
}

<div
  class=" emotion-0"
>
  A
</div>
`;

exports[`Keyboard render default 1`] = `
.emotion-1 {
  position: relative;
  display: inline;
  padding: 3px 4px 6px;
  background: #cbccd2;
  width: auto;
  border-radius: 6px;
  margin: 0;
}

.emotion-0 {
  position: relative;
  display: inline;
  padding: 2px 5px;
  background-image: linear-gradient(rgb(223,224,231),rgb(248,249,255));
  border-radius: 4px;
  box-shadow: 0px 0px 1px 1px rgba(112,113,117,0.8);
  font-size: 12px;
  line-height: 12px;
  color: rgb(152,153,157);
}

.emotion-0 + .emotion-0 {
  margin-left: 18px;
}

.emotion-0 + .emotion-0:before {
  content: '+';
  position: absolute;
  left: -18px;
  top: 0;
  padding: 2px 4px;
  color: rgb(112,113,117);
  font-size: 16px;
}

<ul
  class=" emotion-1"
>
  <li
    class=" emotion-0"
  >
    A
  </li>
</ul>
`;

exports[`Keyboard render with Icon 1`] = `
.emotion-4 {
  position: relative;
  display: inline;
  padding: 3px 4px 6px;
  background: #cbccd2;
  width: auto;
  border-radius: 6px;
  margin: 0;
}

.emotion-0 {
  position: relative;
  display: inline;
  padding: 2px 5px;
  background-image: linear-gradient(rgb(223,224,231),rgb(248,249,255));
  border-radius: 4px;
  box-shadow: 0px 0px 1px 1px rgba(112,113,117,0.8);
  font-size: 12px;
  line-height: 12px;
  color: rgb(152,153,157);
}

.emotion-0 + .emotion-0 {
  margin-left: 18px;
}

.emotion-0 + .emotion-0:before {
  content: '+';
  position: absolute;
  left: -18px;
  top: 0;
  padding: 2px 4px;
  color: rgb(112,113,117);
  font-size: 16px;
}

.emotion-2 {
  position: relative;
  display: inline;
}

.emotion-1 {
  font-size: inherit;
}

<ul
  class="test emotion-4"
>
  <li
    class=" emotion-0"
  >
    A
  </li>
  <li
    class="test emotion-0"
  >
    <span
      class=" emotion-2"
    >
      <i
        class="material-icons emotion-1"
      >
        arrow_upward
      </i>
    </span>
  </li>
</ul>
`;

exports[`Keyboard render with className 1`] = `
.emotion-1 {
  position: relative;
  display: inline;
  padding: 3px 4px 6px;
  background: #cbccd2;
  width: auto;
  border-radius: 6px;
  margin: 0;
}

.emotion-0 {
  position: relative;
  display: inline;
  padding: 2px 5px;
  background-image: linear-gradient(rgb(223,224,231),rgb(248,249,255));
  border-radius: 4px;
  box-shadow: 0px 0px 1px 1px rgba(112,113,117,0.8);
  font-size: 12px;
  line-height: 12px;
  color: rgb(152,153,157);
}

.emotion-0 + .emotion-0 {
  margin-left: 18px;
}

.emotion-0 + .emotion-0:before {
  content: '+';
  position: absolute;
  left: -18px;
  top: 0;
  padding: 2px 4px;
  color: rgb(112,113,117);
  font-size: 16px;
}

<ul
  class="test emotion-1"
>
  <li
    class=" emotion-0"
  >
    A
  </li>
</ul>
`;
